recipes-bsp/u-boot/u-boot-distro-boot/uEnv.txt.in

```diff
@@ -109,6 +112,99 @@ bootcmd_boot=if test ${bootscript_debug} != 1; then \
              fi || true
 
-bootcmd_run=run board_fixups && run check_rollback_needed && run set_bootargs && run set_fdt_path && \
+prog_fuses=if test -n "${fuse_prog_list}" && test "${fuse_status}" = "pending"; then \
+               env set index 1; \
+               for fuse_val in "${fuse_prog_list}"; do \
+                   if test "${fuse_status}" != "failed"; then \
+                       if test "${fuse_dry_run}" = "1"; then \
+                           env set temp echo DRY RUN: fuse prog \\$fuse_bank_word_$index ${fuse_val}; \
+                       else \
+                           env set temp fuse prog -y \\$fuse_bank_word_$index ${fuse_val}; \
+                       fi; \
+                       run temp; \
+                       if test $? != "0"; then \
+                           env set fuse_status failed; \
+                       fi; \
+                       if test "${fuse_dry_run}" = "1"; then \
+                           env set temp echo DRY RUN: fuse cmp \\$fuse_bank_word_$index ${fuse_val}; \
+                       else \
+                           env set temp fuse cmp \\$fuse_bank_word_$index ${fuse_val}; \
+                       fi; \
+                       run temp; \
+                       if test $? != "0"; then \
+                           env set fuse_status failed; \
+                       fi; \
+                   fi ; \
+                   setexpr index $index + 1; \
+                   if test "${index}" = "a"; then \
+                       env set index "10"; \
+                   fi; \
+               done; \
+               @@STATUS_CMD@@; \
+               if test $? != "0"; then \
+                   env set fuse_status failed; \
+               fi; \
+               if test "${fuse_prog_close}" = "1" && test "${fuse_status}" != "failed"; then \
+                   if test "${fuse_dry_run}" = "1"; then \
+                       echo "DRY RUN: Closing is set"; \
+                   else \
+                       @@CLOSE_CMD@@; \
+                   fi; \
+                   if test $? != "0"; then \
+                       env set fuse_status failed; \
+                   fi; \
+               fi; \
+               env set save_vars 1; \
+               if test "${fuse_status}" != "failed"; then \
+                   env set fuse_status "success"; \
+               fi; \
+           fi || true
+
+read_fuses=if test -n "${fuse_num}"; then \
+               for num in "${fuse_num}"; do \
+                   if env exists fuse_val_${num}; then \
+                       env set temp fuse cmp \\$fuse_bank_word_$num 0x\\$fuse_val_$num; \
+                       run temp; \
+                   else \
+                       false; \
+                   fi; \
+                   if test $? != "0"; then \
+                       env set temp fuse readm \\$fuse_bank_word_$num ${loadaddr}; \
+                       run temp; \
+                       setexpr.l fuse_val_${num} "*${loadaddr}"; \
```

Comment:
> Have you tested this code with the hardening enabled. I'd expect to see warning messages saying this command would not be allowed when the device is closed. That is, we'd need to review the command whitelist to ensure everything in this script works both in open and closed states.

No I didn't cause testing every build variant including the various flags in `meta-toradex-security` is quite a lot.

Which command specifically here are you referring to that would cause an issue with the hardening, is it `setexpr.l`? Cause you suggested this command to me in the first place.

> Does the script behave well when HAB is not compiled into the bootloader?

I commented on how the script behaves without HAB/AHAB support on my initial comment on this PR.

> Does the script behave well when the patches you're introducing are not compiled in?

Well without the patches but with HAB/AHAB support, we have the `*_status` commands always returning 0. So the behavior wouldn't be correct in some parts of the script if there are events or something.

The `tdx_is_closed` command would not exist, this would have the same effect as I commented earlier if HAB/AHAB support is not enabled.

> So any code that is not 0 or 16 would mean there's no support for fuse programming.

Well if we think about the cases for `prog_is_closed`. Can't we always assume device is open or device is closed? If `tdx_is_closed` is compiled in then exit code 0 for closed, or 1 for open. If `tdx_is_closed` is not compiled in then that either means HAB/AHAB support is not enabled, in which case the device has to be open or this bootloader wouldn't boot on a closed device without HAB/AHAB enabled.

Or, if `tdx_is_closed` is not available because the bootloader doesn't have these patches for whatever reason but HAB/AHAB is enabled. Then, I guess we can have a 3rd indicator for this, but this kind of defeats the purpose of `read_fuses` right?

The only reason we have `read_fuses` is so we can get the initial values of the fuses to generate an initial yaml file for the fuse package to generate a hash to send to the cloud. If we can't determine/assume the closing state then we can't generate this yaml file properly, which was the whole point of `read_fuses`.

Reply:
> > Have you tested this code...
> 
> No I didn't cause testing every build variant including the various flags in `meta-toradex-security` is quite a lot.

The hardening is enabled by default and this is the configuration we expect users to employ in practice. So, I'd say this would be the most important configuration to test. As I said, I'd expect to see warnings which could be input for us to create a follow up ticket so solve that other part.

> Which command specifically here are you referring to that would cause an issue with the hardening, is it `setexpr.l`? Cause you suggested this command to me in the first place.

Any new commands will most likely not be in the whitelist including `setexpr`.

> > Does the script behave well when HAB is not compiled into the bootloader?
> 
> I commented on how the script behaves without HAB/AHAB support on my initial comment on this PR.

True :crying_cat_face:. Sorry about that.

> > Does the script behave well when the patches you're introducing are not compiled in?
> 
> [...]
> Well if we think about the cases for `prog_is_closed`. Can't we always assume device is open or device is closed?
> [...]

For the reading side, I agree with you. I mean, for real use cases the information reported by `read_fuses` will be correct whether the patches are there or not.

I'm more concerned with the programming side. Do we want users to program the fuses if they are using a bootloader without the patches?

For me it seems they shouldn't and if they were trying it it's a mistake (I mean, I don't see legitimate cases for it).

What I said about the exit code of `tdx_is_closed` was that we could use it to determine if the patches are in and prevent the programming operation. When preventing the operation we could set `fuse_status=no-support` (sorry I wrote `prog_status` before) so that the OS would have a way to know about that and possibly let the platform know why the operation failed.

Reply:
> As I said, I'd expect to see warnings which could be input for us to create a follow up ticket so solve that other part.

I got the following with hardening:

```
## WARNING: Command execution denied (blocked by category) for `fuse readm 0 722...`.
## WARNING: Command execution WOULD BE DENIED in closed state (blocked by category) for `fuse readm 0 722...`.
## WARNING: Command execution WOULD BE DENIED in closed state (blocked by category) for `setexpr.l fuse_val_1 *0x95400000`.
```

So `setexpr` triggers a warning, but `fuse readm` is just outright denied.

> I'm more concerned with the programming side. Do we want users to program the fuses if they are using a bootloader without the patches?

Oh wait you're talking about the programming side? I mean if someone wants to use a non-patched bootloader I don't think that's necessarily wrong right? The patches here aren't explicitly required for programming the fuses they just make the process more robust and safe. The only danger would be that the `*_status` commands now always return 0, meaning we can't prevent devices from being closed in the presence of any events.

This would force users to use the bootloader from whatever version this goes in onwards, or they would need to find these patches in the meta-layer and apply these themselves in whatever custom bootloader they're making. 

If we want to force this limitation we would need to upload updated bootloader packages to the platform after this change goes in so users can easily update to a compatible bootloader. And/or we need to finally publicly document how users can upload their own bootloader binary to use for the bootloader updates.

So I guess it's a matter of do we force this limitation on the feature or not?

In my view I don't see this as a necessity that we have such a limitation. I see it more like a risk assessment choice for the user. If they want to use this, on a older/not-patched U-Boot, then they can but they have to accept that the process is less "safe".

Reply:
> I got the following with hardening:
> [...]

Ok then. I went ahead and created an internal ticket for us to adapt the whitelist/hardening (see TOR-3657) to the new commands in `uEnv.txt` required by this new feature.

> In my view I don't see this as a necessity that we have such a limitation.

If the patches are not strictly required for the programming part to work as you said, then leaving this assessment to the user seems acceptable by me.

Please feel free to resolve this thread after you read this.